PWX-26578: fix the slow task scheduling when runOnce tasks are present

pureneelesh opened the pull request:
**What this PR does / why we need it**:

    In the scheduleTasks function, we iterate over all the tasks to find the ready tasks and
    feed them to the workers. When we encounter a runOnce task, we feed it to the workers and
    then remove it from the task list. But the removal resets the current element's next pointer
    to nil. This terminates the iteration prematurely causing other ready tasks to not get added
    to the work queue.

    Now, we remove the runOnce tasks outside of the iteration.

    Also, added dynamic creation of new worker goroutines.

    - start a new batch (10) of worker goroutines if the number of enqueued tasks
      is high enough

    - excess workers exit after being idle for 30 seconds

    - total number of workers is capped at 100

    Also,

    - fixed staticcheck warnings

    - added some basic stats collection to the scheduler
<!--
  Make sure to have done the following:
  [] Signed off your work as per the DCO.
  [] Add unit-tests
-->



**Which issue(s) this PR fixes** (optional)
PWX-26578

**Special notes for your reviewer**:
Testing:
Modified one of the unit tests. It now fails with the original code.
```
--- FAIL: TestMulti (3.00s)
    /root/git/go/src/github.com/libopenstorage/openstorage/pkg/sched/sched_test.go:106: 
        	Error Trace:	sched_test.go:106
        	Error:      	Should be true
        	Test:       	TestMulti
        	Messages:   	count for runOnce task 3 was not updated
FAIL
FAIL	github.com/libopenstorage/openstorage/pkg/sched	3.004s
FAIL
```

Also, repro'ed the slow scheduling problem locally by creating a lot of mountPath removal tasks. The tasks were running late by as much as 6 minutes. The issue was not seen after the fix.

sangleganesh reviewed (commented):
thanks for the fix. Should we consider dynamically adjusting the number of go routines instead of adding the log ? The number 10 was just arbitrary when we came up with it. 

Also, logs may become too frequent since this is per second. It may be better to log once every 10 minutes.

pureneelesh commented:
> thanks for the fix. Should we consider dynamically adjusting the number of go routines instead of adding the log ? The number 10 was just arbitrary when we came up with it.
> 
Added dynamic increase/decrease of worker threads.

> Also, logs may become too frequent since this is per second. It may be better to log once every 10 minutes.
Removed the log lines and added some basic stats collection and periodic logging instead.

ggriffiths reviewed (approved):
Looks good! Nice change

sangleganesh commented on pkg/sched/sched.go:
```diff
@@ -3,23 +3,96 @@ package sched
+	getHistogram() string
+}
+
+type histImpl struct {
```
nit: add comments for members

pureneelesh replied:
added

sangleganesh commented on pkg/sched/sched.go:
```diff
@@ -3,23 +3,96 @@ package sched
+}
+
+type histImpl struct {
+	lock       sync.Mutex
```
can use composition for lock ?

sangleganesh commented on pkg/sched/sched.go:
```diff
@@ -149,38 +236,90 @@ func (s *manager) Start() {
 			}
-			s.cv.Broadcast()
-			s.enqueuedTasksLock.Unlock()
+			histLastPrinted = time.Now()
```
can use now from line 242 /

sangleganesh commented on pkg/sched/sched.go:
```diff
@@ -149,38 +236,90 @@ func (s *manager) Start() {
+func (s *manager) getWorkerCount() int {
+	var ret int
+	s.Lock()
+	ret = int(s.workersStarted - s.workersExited)
```
nit: can simply be `return int(s.workersStarted - s.workersExited)` with lock and defer ?

sangleganesh reviewed (commented):
still reviewing ...

sangleganesh reviewed (approved):
Looks great, thanks for fixing.
only minor suggestion would be to reduce number of works slowly since we already paid the cost of creating them, but that can be based on your observations on how this is working out.

pureneelesh commented:
> only minor suggestion would be to reduce number of works slowly since we already paid the cost of creating them, but that can be based on your observations on how this is working out.

Added a comment about this. But left the idleTimeout at 30 seconds for now. In my testing, 10 workers seem to be enough after another fix to avoid the duplicate mountPath removal tasks.
